text/0250-Composite_Samplers.md

```diff
@@ -0,0 +1,308 @@
+
+This composite sampler re-uses the concept of Predicates from Approach One.
+
+#### Required Arguments for ConsistentRuleBased
+
+- `SpanKind`
+- list of pairs (`Predicate`, `ConsistentProbabilitySampler`)
+
+For calculating the rejection THRESHOLD, if the `Span` kind matches the specified kind, the sampler goes through the list in the provided order and calls `SpanMatches` on `Predicate`s passing the same arguments as received. If a call returns `true`, the result is as returned by `GetThreshold` called on the corresponding `ConsistentProbabilitySampler`. If the `SpanKind` does not match, or none of the calls to `SpanMatches` yield `true`, the result is obtained by calling `GetThreshold` on `ConsistentAlwaysOffSampler`.
+
+### ConsistentAnyOf
+
+`ConsistentAnyOf` is a composite sampler which takes a non-empty list of ConsistentProbabilitySamplers (delegates) as the argument. The intention is to make a positive sampling decision if __any of__ the delegates would make a positive decision.
+
+Upon invocation of its `GetThreshold` function, it MUST go through the whole list and invoke `GetTheshold` function on each delegate sampler, passing the same arguments as received.
+
+`ConsistentAnyOf` sampler MUST return a THRESHOLD which is constructed as follows:
+
+- If any of the delegates returned a threshold value from the range of `0` to `2^56-1`, the resulting threshold is the minimum value from the set of results from within that range.
+- Otherwise, the result is obtained by calling `GetThreshold` on `ConsistentAlwaysOffSampler`.
+
+Each delegate sampler MUST be given a chance to participate in calculating the threshold as described above and MUST see the same argument values. The order of the delegate samplers does not matter.
+
+### ConsistentRateLimiting
+
+`ConsistentRateLimiting` is a composite sampler that helps control the average rate of sampled spans while allowing another sampler (the delegate) to provide sampling hints.
+
+#### Required Arguments for ConsistentRateLimiting
+
+- ConsistentProbabilitySampler (delegate)
+- maximum sampling (throughput) target rate
+
+The sampler SHOULD measure and keep the average rate of incoming spans, and therefore also of the desired ratio between the incoming span rate to the target span rate.
+Upon invocation of its `GetThreshold` function, the composite sampler MUST get the threshold from the delegate sampler, passing the same arguments as received.
+If using the obtained threshold value as the final threshold would entail sampling more spans than the declared target rate, the sampler SHOULD increase the threshold to a value that would meet the target rate. Several algorithms can be used for threshold adjustment, no particular behavior is prescried by the specification though.
+
+When using `ConsistentRateLimiting` in our requirements example as a replacement for `EachOf` and `RateLimiting`, we are left with no use case for any direct equivalent to `EachOf` in Approach Two.
```

Comment:
I see --
would you agree that the concept of a `ConsistentEachOf` sampler can be defined easily, and that it reduces to the maximum-threshold of the delegates. It's well-defined, but not necessary for the example being developed, I think.

Reply:
Yes. I do not see any difficulties in defining the behavior of `ConsistentEachOf`, it is just its usefulness that is questionable. As soon as we find a nice use case for it, it can be added to this spec.